### PR TITLE
fix(poststate): state clear eip

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Issue

After recent `PostState` fixes, we degraded state root calculation pre EIP 161 (state clear EIP). This also resulted in executor keeping around empty accounts which were touched after Spurious Dragon.

## Solution

Add complete support for account creation/destruction both before and after state clear EIP.

### CODE REVIEW

**Review by onbjerg — APPROVED** [review]

File: crates/revm/src/executor.rs [hunk 1]
```diff
@@ -431,35 +431,48 @@ pub fn commit_state_changes<DB>(
...
-                    if !(has_state_clear_eip && account.is_empty()) {
-                        post_state.create_account(block_number, address, account);
+                    entry.account_state = AccountState::NotExisting; // we will promote account state down the road
+                    let new_account = to_reth_acc(&entry.info);
+
+                    #[allow(clippy::nonminimal_bool)]
+                    // If account was touched before state clear EIP, create it.
+                    if !has_state_clear_eip ||
+                        // If account was touched after state clear EIP, create it only if it is not empty.
+                        (has_state_clear_eip && !new_account.is_empty())
```
Comment (rakita) [inline crates/revm/src/executor.rs:441]: If before state clear, create account.
If after state clear and account is empty, do nothing.

After state clear EIP, only returned empty account from db are the ones that were touched (and are marked for removal) so I a not sure if this is triggered but either way check is good.

File: crates/revm/src/executor.rs [hunk 2]
```diff
@@ -431,35 +431,48 @@ pub fn commit_state_changes<DB>(
...
+
+                    let account_non_existent =
+                        matches!(entry.account_state, AccountState::NotExisting);
+
+                    // Before state clear EIP, create account if it doesn't exist
+                    if (!has_state_clear_eip && account_non_existent)
+                        // After state clear EIP, create account only if it is not empty
+                        || (has_state_clear_eip && entry.info.is_empty() && !new_account.is_empty())
+                    {
+                        post_state.create_account(block_number, address, new_account);
```
Comment (rakita) [inline crates/revm/src/executor.rs:462]: Ok, if the account does not exist or account is empty before state clear, we are creating new account, makes sense.

File: crates/revm/src/executor.rs [hunk 3]
```diff
@@ -474,6 +487,11 @@ pub fn commit_state_changes<DB>(
                 // the account already exists and its storage was cleared, preserve its previous
                 // state
                 AccountState::StorageCleared
+            } else if has_state_clear_eip &&
+                matches!(cached_account.account_state, AccountState::NotExisting) &&
```
Comment (rakita) [inline crates/revm/src/executor.rs:491]: Does this do anything, we are checking if `cached_account.account_state == AccountState::NotExisting` and then setting it to `NotExisting`?

Comment (rkrasiuk) [reply crates/revm/src/executor.rs:491]: more precisely, important that we do not destroy just any empty account, only the ones that were indeed created before state clear EIP

File: crates/revm/src/executor.rs [hunk 4]
```diff
@@ -431,35 +431,48 @@ pub fn commit_state_changes<DB>(
...
+                    } else if old_account != new_account {
                         post_state.change_account(
                             block_number,
                             address,
                             to_reth_acc(&entry.info),
-                            to_reth_acc(&account.info),
+                            new_account,
                         );
-                    } else if has_state_clear_eip && account.is_empty() {
+                    } else if has_state_clear_eip && new_account.is_empty() && !account_non_existent
```
Comment (rakita) [inline crates/revm/src/executor.rs:470]: Yes this is better check, if it is state clear and the incoming account is empty, delete it.

**Review by rakita — APPROVED** [review]
one check is maybe redundant, but the rest of them are okay.

**Review by rakita — APPROVED** [review]
one check is maybe redundant, but the rest of them are okay.

File: crates/revm/src/executor.rs (hunk 3 above)
Comment (rakita) [reply crates/revm/src/executor.rs:491]: But logic does nothing, what it does is:
```
cached_acccount.account_state = if cached_acccount.account_state == AccountState::NotExisting {
     AccountState::NotExisting
}
```

Comment (rakita) [reply crates/revm/src/executor.rs:491]: Aha, you are short `circuit` the else branch

Comment (rkrasiuk) [reply crates/revm/src/executor.rs:491]: the desired behavior is covered in `test_touched_unchanged_account` test. you can comment out this logic and run that test

Comment (rakita) [reply crates/revm/src/executor.rs:491]: It makes sense, I was just looking at `else if` logic and miss `else`

**Review by gakonst — APPROVED** [review]
Nothing to add. Good work on the tests and docs.

**Pull request merged**